core-backend/wasmi/src/memory.rs

```diff
@@ -0,0 +1,136 @@
+// This file is part of Gear.
+
+// Copyright (C) 2022 Gear Technologies Inc.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! wasmi extensions for memory.
+
+use gear_core::memory::{Error, HostPointer, Memory, PageNumber, WasmPageNumber};
+use wasmi::{memory_units::Pages, MemoryRef};
+
+/// Wrapper for [`wasmi::MemoryRef`].
+pub struct MemoryWrap(MemoryRef);
+
+impl MemoryWrap {
+    /// Wrap [`wasmi::MemoryRef`] for Memory trait.
+    pub fn new(mem: MemoryRef) -> Self {
+        MemoryWrap(mem)
+    }
+}
+
+/// Memory interface for the allocator.
+impl Memory for MemoryWrap {
+    fn grow(&mut self, pages: WasmPageNumber) -> Result<PageNumber, Error> {
+        self.0
+            .grow(Pages(pages.0 as usize))
+            .map(|prev| (prev.0 as u32).into())
+            .map_err(|_| Error::OutOfBounds)
+    }
+
+    fn size(&self) -> WasmPageNumber {
+        (self.0.current_size().0 as u32).into()
+    }
+
+    fn write(&mut self, offset: usize, buffer: &[u8]) -> Result<(), Error> {
+        self.0
+            .set(offset as u32, buffer)
+            .map_err(|_| Error::MemoryAccessError)
+    }
+
+    fn read(&self, offset: usize, buffer: &mut [u8]) -> Result<(), Error> {
+        self.0
+            .get_into(offset as u32, buffer)
+            .map_err(|_| Error::MemoryAccessError)
+    }
+
+    fn data_size(&self) -> usize {
+        self.0.current_size().0 * WasmPageNumber::size()
+    }
+
+    unsafe fn get_buffer_host_addr_unsafe(&self) -> HostPointer {
+        self.0.direct_access_mut().as_mut().as_mut_ptr() as HostPointer
+    }
+}
+
+/// can't be tested outside the node runtime
```

Comment:
sorry but didn't get why the node runtime is required there.

Reply:
mistake